Varun Magotra

BE COMPS

2018130022

**Tut 2:** Feature Engineering: Dimensional Reduction

##Questions:<br>
####**Q1. What is ‘Curse of Dimensionality’? Explain**<br>

Curse of dimensionality was first mentioned in the book "Dynamic Programming" by mathmacian R. Bellman. It means that the error increases as the number of features increase. This is pertinent to the complexity of algorithms needed to tackle high dimensionality as well as the exponential increase in running time. Even though higher dimansionality means more information it also means possibility of more noise and redundancy. The difficulty in analysing high-dimensional data results from the conjunction of two effects.

High-dimensional spaces have geometrical properties which are counter-intuitive, and far from the properties which can be observed in two-or three-dimensional spaces.
Data analysis tools are most often designed having in mind intuitive properties and examples in low-dimensional spaces and usually, data analysis tools are best illustrated in 2-or 3-dimensional spaces.
The difficulty is that those tools are also used when data are high-dimensional and more complex and hence, there is a probability to lose the intuition of the behavior of the tool and thus drawing incorrect conclusions.

####**Q2. What is feature selection? Why is it needed? What are the different approaches of feature selection?**<br>
Feature selection means selecting only the features that are not redundant, relevant and consistent. Using proper methods of selection we can improve performance while reducing the running time.

The main benefits of performing feature selection in advance, rather than letting the machine learning model figure out which features are most important, include:

1. **Simpler models:** simple models are easy to explain - a model that is too complex and unexplainable is not valuable
2. **Shorter training times:** a more precise subset of features decreases the amount of time needed to train a model
3. **Variance reduction:** increase the precision of the estimates that can be obtained for a given simulation 
4. **Avoid the curse of high dimensionality:** dimensionally cursed phenomena states that, as dimensionality and the number of features increases, the volume of space increases so fast that the available data become limited - PCA feature selection may be used to reduce dimensionality 

Methods:
1. **Filter methods:** This uses statistics to get the features to be selected instead of cross validation. A fixed metric is used to choose features recursively.
2. **Wrapper methods:** After selecting a set of features they are evaluated and evaluating various combinations we decide which is best. This allows to see the interaction of the features.
3. **Embedded methods:** Embedded feature selection methods integrate the feature selection machine learning algorithm as part of the learning algorithm, in which classification and feature selection are performed simultaneously.




##Problem 1<br>
<p>
Linear Discriminant Analysis, or LDA for short, is a classification machine learning algorithm.

It works by calculating summary statistics for the input features by class label, such as the mean and standard deviation. These statistics represent the model learned from the training data. In practice, linear algebra operations are used to calculate the required quantities efficiently via matrix decomposition.

Predictions are made by estimating the probability that a new example belongs to each class label based on the values of each input feature. The class that results in the largest probability is then assigned to the example.

LDA assumes that the input variables are numeric and normally distributed and that they have the same variance (spread). If this is not the case, it may be desirable to transform the data to have a Gaussian distribution and standardize or normalize the data prior to modeling.

The LDA model is naturally multi-class. This means that it supports two-class classification problems and extends to more than two classes (multi-class classification) without modification or augmentation.

It is a linear classification algorithm, like logistic regression. This means that classes are separated in the feature space by lines or hyperplanes. Extensions of the method can be used that allow other shapes, like Quadratic Discriminant Analysis (QDA), which allows curved shapes in the decision boundary.

We can use PCA to calculate a projection of a dataset and select a number of dimensions or principal components of the projection to use as input to a model.

The scikit-learn library provides the PCA class that can be fit on a dataset and used to transform a training dataset and any additional dataset in the future.

The Linear Discriminant Analysis is available in the scikit-learn Python machine learning library via the LinearDiscriminantAnalysis class.
</p>

##Demonstrate the Linear Discriminant Analysis method with a worked example.
<p>
First, define a synthetic classification dataset or any online dataset you preferred

Note in following information simply based on sytheitic data to explain you the steps of LDA application.

Use the make_classification() function to create a dataset with 1,000 examples, each with 10 input variables.

Following example creates and summarizes the dataset
</p>

### Evaluate a LDA model on the dataset

In [ ]:
from numpy import mean
from numpy import std
from sklearn.datasets import make_classification
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

###Define Dataset

In [ ]:
X, y = make_classification(n_samples=1000, n_features=10, n_informative=10, n_redundant=0, random_state=3)

In [ ]:
print(X.shape, y.shape)

(1000, 10) (1000,)


###Define Model

In [ ]:
model = LinearDiscriminantAnalysis()

###Define model evaluation method

In [ ]:
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

###Evaluate model
<p>
Fit and evaluate a Linear Discriminant Analysis model. You can fit and evaluate a Linear Discriminant Analysis model using repeated stratified k-fold cross-validation via the RepeatedStratifiedKFold class. You will use suitable folds (e.g. 5 or more) and repeat.</p>

In [ ]:
scores = cross_val_score(model, X, y, scoring='accuracy', cv=cv, n_jobs=-1)

###Summarize result

In [ ]:
print('Mean Accuracy: %.3f (%.3f)' % (mean(scores), std(scores)))

Mean Accuracy: 0.833 (0.031)


Running above the example evaluates the Linear Discriminant Analysis algorithm on the synthetic dataset (that I have considered) and reports the average accuracy across the three repeats of 10-fold cross-validation.

I ran the model a few times to see the results.

###Recording results<br>
####<u>Run 1</u>
Mean Accuracy: 0.893 (std: 0.033)
####<u>Run 2</u>
Mean Accuracy: 0.829 (std: 0.031)
####<u>Run 3</u>
Mean Accuracy: 0.833 (std: 0.031)


In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [ ]:
model.fit(X_train, y_train)

LinearDiscriminantAnalysis()

In [ ]:
model.score(X_train, y_train)

0.8253731343283582

In [ ]:
model.score(X_test, y_test)

0.8606060606060606

##Conclusion:
On running LDA muliple times the accuracy changes slightly. We can see that the accuracy is fairly high for test set.